source/MaterialXCore/Types.h

```diff
@@ -115,6 +220,256 @@ class Color4 : public Vector4
     using Vector4::Vector4;
 };
 
+/// The base class for matrices of scalar values
+class MatrixBase { };
+
+/// The class template for matrices of scalar values
+///
+/// Template parameter V specifies the row vector type, and N specifies the
+/// number of row vector in the matrix.
+template <class V, size_t N> class MatrixN : public MatrixBase
+{
+  public:
+    MatrixN() : data{} { }
+    explicit MatrixN(typename V::ScalarType s) { data.fill(V(s)); }
+
+    using iterator = typename std::array<V, N>::iterator;
+    using const_iterator = typename std::array<V, N>::const_iterator;
+
+    /// @name Equality operators
+    /// @{
+
+    /// Return true if the given matrix is identical to this one.
+    bool operator==(const MatrixN& rhs) const { return data == rhs.data; }
+
+    /// Return true if the given vector differs from this one.
+    bool operator!=(const MatrixN& rhs) const { return data != rhs.data; }
+
+    /// @}
+    /// @name Indexing operators
+    /// @{
+
+    /// Return the vector value at the given index.
+    V operator[](size_t i) const { return data.at(i); }
+
+    /// Return the vector value at the given index as a reference.
+    V& operator[](size_t i) { return data.at(i); }
+
+    /// @}
+    /// @name Component-wise operators
+    /// @{
+
+    /// Component-wise addition, returning a new matrix.
+    MatrixN operator+(const MatrixN& rhs) const
+    {
+        MatrixN res;
+        for (size_t i = 0; i < N; i++)
+            res[i] = data[i] + rhs[i];
+        return res;
+    }
+
+    /// Component-wise addition, modifying this matrix in place.
+    MatrixN& operator+=(const MatrixN& rhs)
+    {
+        for (size_t i = 0; i < N; i++)
+            data[i] += rhs[i];
+        return *this;
+    }
+
+    /// Component-wise subtraction, returning a new matrix.
+    MatrixN operator-(const MatrixN& rhs) const
+    {
+        MatrixN res;
+        for (size_t i = 0; i < N; i++)
+            res[i] = data[i] - rhs[i];
+        return res;
+    }
+
+    /// Component-wise subtraction, modifying this matrix in place.
+    MatrixN& operator-=(const MatrixN& rhs)
+    {
+        for (size_t i = 0; i < N; i++)
+            data[i] -= rhs[i];
+        return *this;
+    }
+
+    /// Matrix multplication, returning a new matrix.
+    MatrixN operator*(const MatrixN& rhs) const
+    {
+        static_assert(numRows() == numColumns(), "Requires a square matrix");
+        MatrixN res;
+        for (size_t i = 0; i < N; i++)
+            for (size_t j = 0; j < N; j++)
+                for (size_t k = 0; k < N; k++)
+                    res[i][j] += data[i][k] * rhs[k][j];
+        return res;
+    }
+
+    /// Matrix transpose, returning a new matrix.
```

Comment:
Transpose and translate I've migrated in from previous code in the ViewHandler class.